File: hostel-management-system/src/main/java/t3h/hostelmanagementsystem/controller/utility/UtilityController.java
package t3h.hostelmanagementsystem.controller.utility;

import org.springframework.data.domain.Sort;
import org.springframework.web.bind.annotation.*;
import t3h.hostelmanagementsystem.dto.request.UtilityDTO;
import t3h.hostelmanagementsystem.dto.response.ApiResponse;
import t3h.hostelmanagementsystem.service.utility.UtilityService;

import java.util.List;

@CrossOrigin("*")
@RestController
@RequestMapping("/manager")
public class UtilityController {
    private final UtilityService utilityService;
    public UtilityController(UtilityService utilityService) {
        this.utilityService = utilityService;
    }

    @GetMapping("/get-all-utility")
    public ApiResponse<List<UtilityDTO>> getAllUtility(
            @RequestParam(required = false) String search,
            @RequestParam(required = false, defaultValue = "name") String sort,
            @RequestParam(required = false, defaultValue = "ASC") String direction
    ) {
        // Optional: Implement sorting and searching logic here if needed
        Sort.Direction sortDirection = Sort.Direction.fromString(direction.toUpperCase());
        Sort sortBy = Sort.by(sortDirection, sort);
        // Implement sorting and searching logic here if needed
        ApiResponse<List<UtilityDTO>> apiResponse = new ApiResponse<>();
        apiResponse.setResult(utilityService.getAllUtility(search, sortBy));
        return apiResponse;
    }

    @GetMapping("/get-utility-by-status/{status}")
    public ApiResponse<List<UtilityDTO>> getUtilityByStatus(@PathVariable Integer status) {
        ApiResponse<List<UtilityDTO>> apiResponse = new ApiResponse<>();
        apiResponse.setResult(utilityService.getUtilityByStatus(status));
        return apiResponse;
    }

    @PostMapping("/create-utility")
    public ApiResponse<UtilityDTO> createUtility(@RequestBody UtilityDTO utilityDTO) {
        ApiResponse<UtilityDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(utilityService.addUtility(utilityDTO));
        return apiResponse;
    }

    @PutMapping("/update-utility/{utilityId}")
    public ApiResponse<UtilityDTO> updateUtility(@PathVariable Long utilityId, @RequestBody UtilityDTO utilityDTO) {
        ApiResponse<UtilityDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(utilityService.updateUtility(utilityId, utilityDTO));
        return apiResponse;
    }

    @GetMapping("/get-utility-by-id/{utilityId}")
    public ApiResponse<UtilityDTO> getUtilityById(@PathVariable Long utilityId) {
        ApiResponse<UtilityDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(utilityService.getUtilityById(utilityId));
        return apiResponse;
    }
}

File: fornt-end/frontend-hostelmanagement/src/controller/UtilityController.js
import axios from "axios"
import { UTILITY_API } from "../api/api"
import { useDataByUrl } from "../fetch/FetchData"

export const useGetAllUtility = (filter) => {
    return useDataByUrl({
        url: UTILITY_API.GET_ALL(filter?.search.trim().replace(/\s+/g, ' '), filter?.sort, filter?.direction),
        key: `utilities_search${filter?.search}_sort${filter?.sort}_direction${filter?.direction}`
    })
}

export const useGetAllUtilityByStatus = (status) => {
    return useDataByUrl({
        url: UTILITY_API.GET_UTILITY_BY_STATUS(status),
        key: `utilities_status${status}`
    })
}

export const createUtility = async (data) => {
    try {
        const response = await axios.post(UTILITY_API.CREATE_UTILITY, { ...data });
        return response.data;
    } catch (error) {
        throw error?.response?.data || "Error create utility!";
    }
}

export const updateUtility = async (utilityId, data) => {
    try {
        const response = await axios.put(UTILITY_API.UPDATE_UTILITY(utilityId), { ...data });
        return response.data;
    } catch (error) {
        throw error?.response?.data || "Error update utility!";
    }
}
export const useGetUtilityById = (utilityId) => {
    const enabled = !!utilityId;
    return useDataByUrl({
        url: enabled ? UTILITY_API.GET_UTILITY_BY_ID(utilityId) : null,
        key: `utility_id${utilityId}`,
        enabled,
    })
}
